// fortnox-react/src/autocomplete/Cell.jsx
import React from 'react';
import PropTypes from 'prop-types';

class AutocompleteTableCell extends React.Component {

	getCellStyle() {
		return {
			backgroundColor: this.props.isActive ? '#49c000' : '#FFFFFF'
		};
	}

	getSpanStyle() {
		return {
			color: this.props.isActive ? '#fff' : null
		};
	}

	/**
	 * @memberOf AutocompleteTableCell
	 * @private
	 *
	 * @description
	 * Render Table Cell
	 * The reason why we dont use a css class on the Row for isActive style (background) is
	 * that in IE11 for some reason the DOM wont reflow even when changing classNames in this
	 * particular case. If you set both background-color and color directly on the td Cell, the cells
	 * does not reflow in IE11, hence we use a span with the actual text color.
	 *
	 * @return {Object}
	 */
	render() {
		return (
			<td style={this.getCellStyle()}>
				<span style={this.getSpanStyle()}>{this.props.children}</span>
			</td>
		);
	}

}

AutocompleteTableCell.defaultProps = {
	isActive: false
};

AutocompleteTableCell.propTypes = {
	isActive: PropTypes.bool,
	children: PropTypes.node.isRequired
};

export default AutocompleteTableCell;
