// fortnox-react/src/table/Filter.jsx
import React from 'react';
import PropTypes from 'prop-types';

class Filter extends React.Component {

	/**
	 * @private
	 * @memberOf Filter
	 *
	 * @description
	 * Call parent click handler
	 *
	 * @returns {undefined}
	 */
	handleClick() {
		this.props.onClick(this.props.filterID);
	}

	/**
	 * @private
	 * @memberOf Filter
	 *
	 * @description
	 * Generate list of classes
	 *
	 * @returns {string}
	 */
	getCSSClasses() {
		return [
			'filter',
			this.props.color,
			this.props.isActive ? 'active' : ''
		].join(' ');
	}

	/**
	 * @private
	 * @memberOf Filter
	 *
	 * @description
	 * Render the filter
	 *
	 * @return {Object} React Elements
	 */
	render() {
		return (
			<span className={this.getCSSClasses()} onClick={this.handleClick.bind(this)}>
				{this.props.children}
			</span>
		);
	}

}

Filter.defaultProps = {
	isActive: false,
	onClick() { return null; }
};


Filter.propTypes = {
	color: PropTypes.string.isRequired,
	filterID: PropTypes.string.isRequired,
	onClick: PropTypes.func.isRequired,
	isActive: PropTypes.bool,
	children: PropTypes.node.isRequired
};


export default Filter;
