// fortnox-react/src/flexbox/Container.jsx
import React from 'react';
import PropTypes from 'prop-types';

/**
 * FlexContainer that when used adds display
 * flex and the possibility of aligning the children
 * both horizontally and vertically.
 *
 * @class
 * @name FlexContainer
 *
 * @return {FlexContainer}
 */
class FlexContainer extends React.Component {

	contains(node) {
		return this.el.contains(node);
	}

	/**
	 * @memberOf FlexContainer
	 * @private
	 *
	 * @description
	 * Default menu is not visible
	 *
	 * @return {String} list of classes to use
	 */
	getClasses() {
		return [
			this.props.className,
			this.props.align ? `flex-align-${this.props.align}` : null
		].filter(cls => cls !== null).join(' ');
	}

	/**
	 * @memberOf FlexContainer
	 * @private
	 *
	 * @description
	 * Render the Container
	 *
	 * @returns {Object} React Elements
	 */
	render() {
		return <div ref={(el) => { this.el = el; }} className={`flex-container ${this.getClasses()}`}>{this.props.children}</div>;
	}

}

FlexContainer.defaultProps = {
	className: null,
	align: null,
	children: null
};

FlexContainer.propTypes = {
	className: PropTypes.string,
	align: PropTypes.string,
	children: PropTypes.node
};

export default FlexContainer;
