// fortnox-react/src/flexbox/Row.jsx
import React from 'react';
import PropTypes from 'prop-types';

/**
 * The React Row component.
 *
 *
 * @class
 * @name Row
 *
 * @return {Row} An instance of the Row component.
 */
class Row extends React.Component {

	getAlignment() {
		return {
			justifyContent: this.props.alignMap[this.props.align],
			WebkitJustifyContent: this.props.alignMap[this.props.align]
		};
	}

	/**
	 * @private
	 * @memberOf Row
	 *
	 * @description
	 * Return a list of classes to be used on the Row
	 *
	 * @return {String}
	 */
	getCSSClasses() {
		return `flex-row ${this.props.className}`;
	}

	/**
	 * @private
	 * @memberOf Row
	 *
	 * @description
	 * Renders the nested flex Row with children
	 *
	 * @return {Object} - React internals
	 */
	render() {
		return (
			<div className={this.getCSSClasses()} style={this.getAlignment()}>
				{this.props.children}
			</div>
		);
	}

}

Row.defaultProps = {
	align: 'default',
	className: '',
	alignMap: {
		default: 'space-between',
		left: 'flex-start',
		right: 'flex-end',
		center: 'center'
	}
};

Row.propTypes = {
	alignMap: PropTypes.object,
	align: PropTypes.string,
	className: PropTypes.string,
	children: PropTypes.node.isRequired
};


export default Row;
