// fortnox-react/src/form/ExpanderContent.jsx
import React from 'react';
import PropTypes from 'prop-types';

/**
 * The React form.ExpanderContent Component
 * Renders the Content of the expander
 *
 * @class
 * @name ExpanderContent
 *
 * @return {ExpanderContent} An instance of the form.ExpanderContent component.
 */
class ExpanderContent extends React.Component {

	/**
	 * @private
	 * @memberOf ExpanderContent
	 *
	 * @description
	 *
	 * Get the styles to use for the form-expander
	 *
	 * @return {Object}
	 */
	getStyle() {
		return {
			height: this.props.isExpanded ? 'auto' : 0,
			overflow: this.props.isExpanded ? 'visible' : 'hidden'
		};
	}

	/**
	 * @private
	 * @memberOf ExpanderContent
	 *
	 * @description
	 *
	 * Render the expander Content
	 *
	 * @return {Object} - React internals
	 */
	render() {
		return (
			<div className="form-expander" style={this.getStyle()}>
				<div className="snitt-shadow" />
				<div className="expander-content">
					{this.props.children}
				</div>
				<div className="end-snitt-shadow" />
			</div>
		);
	}

}

ExpanderContent.propTypes = {
	isExpanded: PropTypes.bool.isRequired,
	children: PropTypes.node.isRequired
};

export default ExpanderContent;
