// should-enzyme/lib/react-enzyme.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['exports', 'enzyme/build/react-compat', './base-enzyme'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('enzyme/build/react-compat'), require('./base-enzyme'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.reactCompat, global.baseEnzyme);
    global.reactEnzyme = mod.exports;
  }
})(this, function (exports, _reactCompat, _baseEnzyme) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _baseEnzyme2 = _interopRequireDefault(_baseEnzyme);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
      throw new TypeError("Cannot call a class as a function");
    }
  }

  var _createClass = function () {
    function defineProperties(target, props) {
      for (var i = 0; i < props.length; i++) {
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
      }
    }

    return function (Constructor, protoProps, staticProps) {
      if (protoProps) defineProperties(Constructor.prototype, protoProps);
      if (staticProps) defineProperties(Constructor, staticProps);
      return Constructor;
    };
  }();

  function _possibleConstructorReturn(self, call) {
    if (!self) {
      throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }

    return call && (typeof call === "object" || typeof call === "function") ? call : self;
  }

  function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
      throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
    }

    subClass.prototype = Object.create(superClass && superClass.prototype, {
      constructor: {
        value: subClass,
        enumerable: false,
        writable: true,
        configurable: true
      }
    });
    if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
  }

  var ReactEnzyme = function (_BaseEnzyme) {
    _inherits(ReactEnzyme, _BaseEnzyme);

    function ReactEnzyme(enzymeWrapper) {
      _classCallCheck(this, ReactEnzyme);

      var _this = _possibleConstructorReturn(this, (ReactEnzyme.__proto__ || Object.getPrototypeOf(ReactEnzyme)).call(this));

      _this.enzyme = enzymeWrapper;
      return _this;
    }

    _createClass(ReactEnzyme, [{
      key: 'classNames',
      value: function classNames() {
        return this.element.className;
      }
    }, {
      key: 'hasClass',
      value: function hasClass(className) {
        var classNameList = this.classNames().split(' ');
        return classNameList.indexOf(className) > -1;
      }
    }, {
      key: 'attr',
      value: function attr(name) {
        return this.element.getAttribute(name) || undefined;
      }
    }, {
      key: 'checked',
      value: function checked() {
        return this.element.checked;
      }
    }, {
      key: 'value',
      value: function value() {
        return this.element.value;
      }
    }, {
      key: 'element',
      get: function get() {
        if (!this.__element) this.__element = this.enzyme.single(function (n) {
          return (0, _reactCompat.findDOMNode)(n);
        });

        return this.__element;
      }
    }]);

    return ReactEnzyme;
  }(_baseEnzyme2.default);

  exports.default = ReactEnzyme;
});
