// should-enzyme/lib/shallow-enzyme.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['exports', './base-enzyme', 'cheerio'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('./base-enzyme'), require('cheerio'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.baseEnzyme, global.cheerio);
    global.shallowEnzyme = mod.exports;
  }
})(this, function (exports, _baseEnzyme, _cheerio) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _baseEnzyme2 = _interopRequireDefault(_baseEnzyme);

  var _cheerio2 = _interopRequireDefault(_cheerio);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
      throw new TypeError("Cannot call a class as a function");
    }
  }

  var _createClass = function () {
    function defineProperties(target, props) {
      for (var i = 0; i < props.length; i++) {
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
      }
    }

    return function (Constructor, protoProps, staticProps) {
      if (protoProps) defineProperties(Constructor.prototype, protoProps);
      if (staticProps) defineProperties(Constructor, staticProps);
      return Constructor;
    };
  }();

  function _possibleConstructorReturn(self, call) {
    if (!self) {
      throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }

    return call && (typeof call === "object" || typeof call === "function") ? call : self;
  }

  function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
      throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
    }

    subClass.prototype = Object.create(superClass && superClass.prototype, {
      constructor: {
        value: subClass,
        enumerable: false,
        writable: true,
        configurable: true
      }
    });
    if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
  }

  var ShallowEnzyme = function (_BaseEnzyme) {
    _inherits(ShallowEnzyme, _BaseEnzyme);

    function ShallowEnzyme(enzymeWrapper) {
      _classCallCheck(this, ShallowEnzyme);

      var _this = _possibleConstructorReturn(this, (ShallowEnzyme.__proto__ || Object.getPrototypeOf(ShallowEnzyme)).call(this));

      _this.enzyme = enzymeWrapper;
      return _this;
    }

    _createClass(ShallowEnzyme, [{
      key: 'classNames',
      value: function classNames() {
        return this.enzyme.prop('className');
      }
    }, {
      key: 'hasClass',
      value: function hasClass(className) {
        return this.enzyme.hasClass(className);
      }
    }, {
      key: 'attr',
      value: function attr(name) {
        return this.element.attr(name);
      }
    }, {
      key: 'checked',
      value: function checked() {
        return this.element.is(':checked');
      }
    }, {
      key: 'element',
      get: function get() {
        if (!this.__element) this.__element = (0, _cheerio2.default)(this.enzyme.html());

        return this.__element;
      }
    }]);

    return ShallowEnzyme;
  }(_baseEnzyme2.default);

  exports.default = ShallowEnzyme;
});
