// fortnox-react/src/utils/ConfirmDialog.jsx
import React from 'react';
import PropTypes from 'prop-types';
import { t as gettext } from 'c-3po';
import Dialog from './Dialog';
import Button from '../form/Button';
import Row from '../flexbox/Row';
import Column from '../flexbox/Column';
import Flex from '../flexbox/Container';

class ConfirmDialog extends React.Component {
	renderDescription() {
		if (this.props.description) {
			return (
				<Row>
					<Column>
						{this.props.description}
					</Column>
				</Row>);
		}
		return null;
	}

	render() {
		return (
			<Dialog
				width="380" height="auto"
				title={this.props.title}
				isOpen={this.props.isOpen} blockViewport
				color={this.props.color}
			>
				{this.renderDescription()}
				{this.props.children}
				<Row>
					<Flex align="center">
						<Button
							color="gray"
							onClick={this.props.onUserAbort}
                            disabled={this.props.disableCancel}
						>
							{this.props.cancelText}
						</Button>
						<Button
							color="green"
							onClick={this.props.onUserConfirm}
                            disabled={this.props.disableOk}
						>
							{this.props.okText}
						</Button>
					</Flex>
				</Row>

			</Dialog>
		);
	}

}

ConfirmDialog.defaultProps = {
	cancelText: gettext`Avbryt`,
	okText: gettext`Ok`,
	isOpen: false,
	color: 'green',
    disableOk: false,
    disableCancel: false
};

ConfirmDialog.propTypes = {
	children: PropTypes.node.isRequired,
	title: PropTypes.string.isRequired,
	cancelText: PropTypes.string,
	okText: PropTypes.string,
	onUserAbort: PropTypes.func.isRequired,
	onUserConfirm: PropTypes.func.isRequired,
	description: PropTypes.string.isRequired,
	isOpen: PropTypes.bool.isRequired,
	color: PropTypes.string.isRequired,
    disableOk: PropTypes.bool,
    disableCancel: PropTypes.bool
};

export default ConfirmDialog;
