// fortnox-react/src/utils/NotificationMessage.jsx
import React from 'react';
import PropTypes from 'prop-types';

class NotificationMessage extends React.Component {
	getIcon() {
		const map = {
			info: 'fa fa-info',
			warning: 'fa fa-exclamation-triangle',
			danger: 'fa fa-exclamation',
			success: 'fa fa-check'
		};
		if (this.props.icon) {
			return <i className={this.props.icon} />;
		}
		if (!map[this.props.type]) {
			return '';
		}

		return <i className={map[this.props.type]} />;
	}

	getDisplay() {
		if (this.props.open) {
			return 'open';
		}
		return 'hidden';
	}

	render() {
		return (
			<div className={`${this.props.className} notificationmessage ${this.props.type} ${this.getDisplay()}`}>
				<div className="symbol">
					{this.getIcon()}
				</div>
				<div className="infomessage">
					{this.props.children}
				</div>
			</div>
		);
	}
}

NotificationMessage.propTypes = {
	type: PropTypes.oneOf(['warning', 'success', 'info', 'danger']),
	icon: PropTypes.string,
	open: PropTypes.bool,
	className: PropTypes.string,
	children: PropTypes.node.isRequired
};

NotificationMessage.defaultProps = {
	type: 'warning',
	icon: null,
	open: true,
	className: ''
};

export default NotificationMessage;
