// fortnox-react/src/view/ViewTool.jsx
import React from 'react';
import PropTypes from 'prop-types';

/**
 * The React ViewToolButton Component
 * Renders tool buttons inside of a view
 *
 * @class
 * @name ViewToolButton
 *
 * @return {ViewToolButton} An instance of the ViewToolButton component.
 */
class ViewToolButton extends React.Component {

	/**
	 * @memberOf ViewToolButton
	 * @private
	 *
	 * @description
	 * Only rerender when we have new props
	 *
	 * @param {Object} nextProps
	 *
	 * @return {Boolean}
	 */
	shouldComponentUpdate(nextProps) {
		return this.props !== nextProps;
	}

	/**
	 * @memberOf ViewToolButton
	 * @private
	 *
	 * @description
	 * Handle clicks on tool buttons
	 *
	 * @returns {undefined}
	 */
	handleClick() {
		this.props.onUserClick();
	}

	/**
	 * @memberOf ViewToolButton
	 * @private
	 *
	 * @description
	 * Handle clicks on tool buttons
	 *
	 * @return {Object}
	 */
	render() {
		return (
			<li>
				<button onClick={this.handleClick.bind(this)} className={this.props.icon}>
					{this.props.title}
				</button>
			</li>
		);
	}
}


ViewToolButton.propTypes = {
	onUserClick: PropTypes.func.isRequired,
	icon: PropTypes.string.isRequired,
	title: PropTypes.string.isRequired
};


export default ViewToolButton;
