// fortnox-react/src/view/ViewTabMenu.jsx
import React from 'react';
import PropTypes from 'prop-types';
import Flex from '../flexbox/Container';
import Registers from '../utils/RegisterDropDown';
import Reports from '../utils/ReportsButton';

/**
 * Renders a tab menu with reports already defined
 *
 * @class
 * @name ViewTabMenu
 *
 * @return {ViewTabMenu}
 */
class ViewTabMenu extends React.Component {
	getClasses() {
		let classes = `view-tab-menu ${this.props.className}`;
		if (window.parent.isVessel) {
			classes = `${classes} tab-fixed`;
		}
		return classes;
	}

	renderReports() {
		if (!window.parent.isVessel) {
			return null;
		}

		return <Reports eventbus={this.props.eventbus} />;
	}

	renderRegisters() {
		if (!window.parent.isVessel) {
			return null;
		}

		return <Registers eventbus={this.props.eventbus} />;
	}

	render() {
		return (
			<div>
				<div className={this.getClasses()}>
					<Flex align="left">
						{this.props.children}
					</Flex>
					<Flex align="right">
						{ this.renderReports() }
						{ this.renderRegisters() }
					</Flex>
				</div>
				<div className="view-tab-bar-placeholder" />
			</div>
		);
	}
}

ViewTabMenu.propTypes = {
	className: PropTypes.string,
	children: PropTypes.node.isRequired,
	eventbus: PropTypes.object.isRequired
};

ViewTabMenu.defaultProps = {
	className: ''
};

export default ViewTabMenu;
